polygeist/lib/Dialect/Polygeist/Analysis/MemoryAccessAnalysis.cpp

```diff
@@ -90,36 +90,250 @@ static bool isSmallerThanNegativeOne(Value val, DataFlowSolver &solver) {
   return (optConstVal.has_value() && optConstVal->slt(-1));
 }
 
-/// Remove conversion like operations in the incoming expression if any.
-static Value removeConversions(Value expr) {
-  if (isa<CastOpInterface>(expr.getDefiningOp()))
-    return expr.getDefiningOp()->getOperand(0);
-  return expr;
+/// Walks up the parents and records the ones with the specified type.
+template <typename T> static SetVector<T> getParentsOfType(Block *block) {
+  SetVector<T> res;
+  constexpr auto getInitialParent = [](Block *b) -> T {
+    Operation *op = b->getParentOp();
+    auto typedOp = dyn_cast<T>(op);
+    return typedOp ? typedOp : op->getParentOfType<T>();
+  };
+
+  for (T parent = getInitialParent(block); parent;
+       parent = parent->template getParentOfType<T>())
+    res.insert(parent);
+
+  return res;
 }
 
-/// Determine whether 'expr' is a multiplication involving 'val'.
-/// TODO: Rewrite this function to handle more generic expressions.
-/// Example:
-///   %0 = arith.index_cast %ii : index to i64
-///   %i = arith.muli %0, %c1 : i64
-/// In this case we fail to match because the %0 is not %ii.
-static std::optional<Value> getMultiplier(Value expr, Value val) {
-  expr = removeConversions(expr);
-  if (expr == val) {
-    OpBuilder b(expr.getContext());
-    return b.create<arith::ConstantIndexOp>(b.getUnknownLoc(), 1);
+/// Determine whether \p op uses \p val (directly or indirectly).
+static bool usesValue(Operation *op, Value val) {
+  return (llvm::any_of(val.getUsers(), [&](Operation *user) {
+    if (user == op)
+      return true;
+    return llvm::any_of(user->getResults(),
+                        [&](Value res) { return usesValue(op, res); });
+  }));
+}
```

Comment:
Do you think it could be faster, and more initiative to go from op to val, then val to op? 
```suggestion
static bool usesValue(Operation *op, Value val) {
  assert(op && "Expecting valid operation");
  if (Operation *valOp = val.getDefiningOp())
    if (valOp == op)
      return true;
  return llvm::any_of(op->getOperands(), [&](Value operand) {
    if (Operation *operandOp = operand.getDefiningOp())
      return usesValue(operandOp, val);
    return (operand == val);
  });
}
```

Reply:
Arguable which one is better, The original uses llvm::any_of twice, the new one uses it once. Probably the new one is easier to understand so I can adopt the suggestion.